09_json/n0npax/pkg/puppy/store/memStore.go

```diff
@@ -0,0 +1,69 @@
+package store
+
+import (
+	"sync"
+
+	puppy "github.com/anz-bank/go-course/09_json/n0npax/pkg/puppy"
+)
+
+// MemStore map based type for storing puppies data
+type MemStore struct {
+	m      map[int]puppy.Puppy
+	nextID int
+	sync.Mutex
+}
+
+// NewMemStore creates new storer for map
+func NewMemStore() *MemStore {
+	return &MemStore{m: map[int]puppy.Puppy{}}
+}
+
+// CreatePuppy creates puppy
+func (m *MemStore) CreatePuppy(p *puppy.Puppy) (int, error) {
+	if p.Value < 0 {
+		return -1, puppy.Errorf(puppy.ErrCodeInvalidInput, "Puppy value have to be positive number")
+	}
+	m.Lock()
+	defer m.Unlock()
+	id := m.nextID
+	m.nextID++
+	p.ID = id
+	m.m[id] = *p
+	return id, nil
+}
+
+// ReadPuppy reads puppy from backend
+func (m *MemStore) ReadPuppy(id int) (*puppy.Puppy, error) {
+	if puppy, ok := m.m[id]; ok {
+		return &puppy, nil
+	}
+	return nil, puppy.Errorf(puppy.ErrCodeNotFound, "puppy with ID (%v) not found", id)
+}
+
+// UpdatePuppy updates puppy
+func (m *MemStore) UpdatePuppy(id int, p *puppy.Puppy) error {
```

Comment:
I'd skip passing the ID separately, as Puppy already has one. If you're updating, the ID shouldn't change, so the ID in Puppy should be enough.

Reply:
This is holly war :)

Example PUT in K8S (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.14/#volumemount-v1-core)
`PUT /apis/batch/v1beta1/namespaces/{namespace}/cronjobs/{name}`

We could extract `{name}` from payload, but it's provided explicitly.

Reply:
:) Indeed can be a matter of taste. I agree with using it explicitly in case of a PUT, but just doesn't feel right in the case of internal language functions. But it's taste anyway :)